rpc: Rename size to vsize in mempool related calls

IPGlider opened the pull request:
In `getmempoolancestors`, `getmempooldescendants`, `getmempoolentry` and `getrawmempool` RPCs `size` returns the virtual transaction size as defined in BIP 141. Renaming it to `vsize` makes it consistent with returned value and other calls such as `getrawtransaction`.

Related to #11218.

IPGlider committed:
rpc: Rename size to vsize in mempool related calls

fanquake commented:
@IPGlider You'll also need to update ```test/functional/mining_prioritisetransaction.py```. See the Travis [build failure](https://travis-ci.org/bitcoin/bitcoin/jobs/367787349#L2408).

IPGlider commented:
@fanquake Thanks for letting me know. I have now fixed the tests. I will squash the commits if requested.

Empact commented:
Squash please.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -383,7 +383,7 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
     AssertLockHeld(mempool.cs);
 
-    info.pushKV("size", (int)e.GetTxSize());
+    info.pushKV("vsize", (int)e.GetTxSize());
```
I imagine this key-value pair is heavily relied upon. I'd prefer if we included both keys "size" and "vsize" with the same value for at least one release.

promag replied:
Agree with @MarcoFalke, this is a breaking change.

If you feel that field `size` should be removed then you must deprecate it first, and wait for the next release to actually remove it. Not sure it's worth it as the description is clear.

IPGlider commented:
I have kept the size field for compatibility as requested by @MarcoFalke. I also added a deprecation message and squashed the commits. Let me know if anything else is needed.

promag commented on doc/release-notes.md:
```diff
@@ -70,6 +70,7 @@ RPC changes
   /rest/block/ endpoints when in json mode. This is also included in `getblock`
   (with verbosity=2), `listsinceblock`, `listtransactions`, and
   `getrawtransaction` RPC commands.
+- In `getmempoolancestors`, `getmempooldescendants`, `getmempoolentry` and `getrawmempool` RPCs `size` has been renamed to `vsize` to be consistent with the returned value and other RPCs such as `getrawtransaction`.
```
Please update note, something like ``In ... `vsize` was added and `size` is now deprecated ...``

jnewbery commented:
Concept ACK. A couple of nitty suggestions:
- return `vsize` before `size` (since the definition for `size` depends on `vsize`)
- hide `size` behind a deprecation flag so it can be removed cleanly in the next version.

Needs rebase.

IPGlider commented:
@jnewbery Changed the order and rebased.

Could you please point me towards an example of how to hide `size` behind a deprecation flag?

promag commented:
@IPGlider See https://github.com/bitcoin/bitcoin/blob/81d0d565ab5551110ef465484b061922452aabe9/src/wallet/rpcwallet.cpp#L1763

IPGlider commented:
Thanks @promag. 

Hidden `size` behind a deprecation flag. Any further suggestions?

laanwj commented:
I'm not sure about this anymore - this came up in some discussions at the building on bitcoin conference. In practice almost no one knows what `vsize` is, and everyone that knows what it is likely already knows that all the sizes are virtual now.

luke-jr commented:
People probably shouldn't use `"size"` if they don't know about vsize, so I think renaming it here still makes sense...

MarcoFalke commented:
utACK 52606162ae1cc3c954de7a37411910784317315a

MarcoFalke commented:
re-utACK 305330e818a410d5bb1f8fc2b6fea0442c3a06a0

Empact commented:
Concept ACK - I'm in favor since we show virtual size in the gui as of at least: #12580

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -351,7 +351,8 @@ static UniValue getdifficulty(const JSONRPCRequest& request)
 {
-    return "    \"size\" : n,             (numeric) virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted.\n"
+    return "    \"vsize\" : n,            (numeric) virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted.\n"
+           "    \"size\" : n,             (numeric) same as vsize (DEPRECATED)\n"
```
I think this should document that `size` is only returned if `deprecatedrpc=size` is in the config or passed as an argument to bitcoind. Perhaps:

```
           "    \"size\" : n,             (numeric) (DEPRECATED) same as vsize. Only returned if bitcoind is started with -deprecatedrpc=size\n"
           "                              `size` will be completely removed in v0.18.\n"
```

jnewbery commented on doc/release-notes.md:
```diff
@@ -81,6 +81,10 @@ RPC changes
+- In `getmempoolancestors`, `getmempooldescendants`, `getmempoolentry` and
+  `getrawmempool` RPCs, to be consistent with the returned value and other RPCs
+  such as `getrawtransaction`, `vsize` has been added and `size` is now
+  deprecated.
```
Should document that `size` will only be returned if bitcoind is started with -deprecatedrpc=size.

jnewbery reviewed (commented):
Looks fine, but documentation could be improved for the deprecatedrpc option.

jnewbery commented:
tACK 3bc922d79c43113ad9face78b2c6208e38c72d47.

Thanks for addressing my review comments. Sorry I didn't ACK earlier. I didn't notice that you'd updated the branch.

MarcoFalke commented:
re-utACK 3bc922d79c

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -198,6 +198,12 @@ RPC
+  getrawmempool RPCs, to be consistent with the returned value and other
+  RPCs such as getrawtransaction, vsize has been added and size is now
+  deprecated. size will only be returned if bitcoind is started with
+  -deprecatedrpc=size.
```
Should be moved to a separate file `./doc/release-notes-13008.md`

fanquake commented:
I've rebased and fixed up the release notes in #15637.
